# CICIDS2017 데이터 EDA

### 1) Dataset Import
- Datafram을 위한 pandas import
- monday, friday 모두 import

In [1]:
import pandas as pd

In [2]:
mon_csv = pd.read_csv("data/raw/Monday-WorkingHours.pcap_ISCX.csv")
fri_csv = pd.read_csv("data/raw/Friday-WorkingHours-Afternoon-DDos.pcap_ISCX.csv")

- 잘 import 되었는지 확인 

In [3]:
mon_csv.head()

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,49188,4,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
1,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
2,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
3,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
4,49486,3,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


In [4]:
fri_csv.head()

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,54865,3,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
1,55054,109,1,1,6,6,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
2,55055,52,1,1,6,6,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
3,46236,34,1,1,6,6,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
4,54863,3,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


- 불필요한 공백 제거

In [5]:
mon_csv.columns = mon_csv.columns.str.strip()
fri_csv.columns = fri_csv.columns.str.strip()

In [6]:
mon_csv.head()

,Destination Port,Flow Duration,Total Fwd Packets,Total Backward Packets,Total Length of Fwd Packets,Total Length of Bwd Packets,Fwd Packet Length Max,Fwd Packet Length Min,Fwd Packet Length Mean,Fwd Packet Length Std,...,min_seg_size_forward,Active Mean,Active Std,Active Max,Active Min,Idle Mean,Idle Std,Idle Max,Idle Min,Label
0,49188,4,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
1,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
2,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
3,49188,1,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN
4,49486,3,2,0,12,0,6,6,6.0,0.0,...,20,0.0,0.0,0,0,0.0,0.0,0,0,BENIGN


In [7]:
print("컬럼 개수 일치:", len(mon_csv.columns) == len(fri_csv.columns))
print("컬럼명 완전 일치:", list(mon_csv.columns) == list(fri_csv.columns))


컬럼 개수 일치: True
컬럼명 완전 일치: True


In [8]:
mon_csv.columns.tolist()

['Destination Port',
 'Flow Duration',
 'Total Fwd Packets',
 'Total Backward Packets',
 'Total Length of Fwd Packets',
 'Total Length of Bwd Packets',
 'Fwd Packet Length Max',
 'Fwd Packet Length Min',
 'Fwd Packet Length Mean',
 'Fwd Packet Length Std',
 'Bwd Packet Length Max',
 'Bwd Packet Length Min',
 'Bwd Packet Length Mean',
 'Bwd Packet Length Std',
 'Flow Bytes/s',
 'Flow Packets/s',
 'Flow IAT Mean',
 'Flow IAT Std',
 'Flow IAT Max',
 'Flow IAT Min',
 'Fwd IAT Total',
 'Fwd IAT Mean',
 'Fwd IAT Std',
 'Fwd IAT Max',
 'Fwd IAT Min',
 'Bwd IAT Total',
 'Bwd IAT Mean',
 'Bwd IAT Std',
 'Bwd IAT Max',
 'Bwd IAT Min',
 'Fwd PSH Flags',
 'Bwd PSH Flags',
 'Fwd URG Flags',
 'Bwd URG Flags',
 'Fwd Header Length',
 'Bwd Header Length',
 'Fwd Packets/s',
 'Bwd Packets/s',
 'Min Packet Length',
 'Max Packet Length',
 'Packet Length Mean',
 'Packet Length Std',
 'Packet Length Variance',
 'FIN Flag Count',
 'SYN Flag Count',
 'RST Flag Count',
 'PSH Flag Count',
 'ACK Flag Count',
 'UR

- 모양 확인 

In [9]:
print("Monday shape:", mon_csv.shape)
print("Friday-DDoS shape:", fri_csv.shape)

Monday shape: (529918, 79)
Friday-DDoS shape: (225745, 79)


- 하나의 결과물로 저장하기 위해 하나로 합침 
- `A_source`라는 컬럼을 새로 만듦 -> 해당 데이터가 어디서 나온건지 확인 

In [10]:
mon_csv["A_source"] = "Monday"
fri_csv["A_source"] = "Friday-DDoS"

In [11]:
df = pd.concat([mon_csv, fri_csv], ignore_index=True)

In [12]:
print("합친 shape:", df.shape)

합친 shape: (755663, 80)


### 2) 데이터 결함 확인

In [13]:
import numpy as np

In [14]:
### 수치형 컬럼 확인
ncol = df.select_dtypes(include=[np.number]).columns

- 결측치 확인 : 결함있는 데이터가 있음 
  - Flow Bytes/s
  - Flow Packets/s -> 471개 
  - Flow Duration = 0 인 행 수: 471
  - 동일하게 471개 -> 뭔가 관련이 있을 것으로 보임

In [15]:
inf_counts = np.isinf(df[ncol]).sum()
print("\n[Inf 있는 컬럼]")
print(inf_counts[inf_counts > 0])

nan_counts = df[ncol].isna().sum()
print("\n[NaN 있는 컬럼]")
print(nan_counts[nan_counts > 0])

dup_full = df.duplicated().sum()
print("\n전체 컬럼(Label 포함) 완전 일치 중복 행 수:", dup_full)

neg_counts = (df[ncol] < 0).sum()
print("\n[음수 있는 컬럼]")
print(neg_counts[neg_counts > 0])

zero_dur = df[df["Flow Duration"] == 0]
print("\nFlow Duration = 0 인 행 수:", len(zero_dur))
print(zero_dur["Label"].value_counts())


[Inf 있는 컬럼]
Flow Bytes/s      403
Flow Packets/s    471
dtype: int64

[NaN 있는 컬럼]
Flow Bytes/s    68
dtype: int64

전체 컬럼(Label 포함) 완전 일치 중복 행 수: 29568

[음수 있는 컬럼]
Flow Duration                  17
Flow Bytes/s                   11
Flow Packets/s                 17
Flow IAT Mean                  17
Flow IAT Max                   17
Flow IAT Min                  620
Fwd IAT Min                     6
Fwd Header Length              23
Bwd Header Length              17
Fwd Header Length.1            23
Init_Win_bytes_forward     257420
Init_Win_bytes_backward    387393
min_seg_size_forward           23
dtype: int64

Flow Duration = 0 인 행 수: 471
Label
BENIGN    469
DDoS        2
Name: count, dtype: int64


- 음수로 나온 컬럼 확인 

In [16]:
print("Header Length 음수값 분포:")
print(df.loc[df["Fwd Header Length"] < 0, "Fwd Header Length"].describe())

print("\nDuration 음수(17건)와 Packets/s·Bytes/s 음수 겹침:")
neg_dur = df["Flow Duration"] < 0
print("Duration & Packets/s 동시:", (neg_dur & (df["Flow Packets/s"] < 0)).sum())
print("Duration & Bytes/s 동시:", (neg_dur & (df["Flow Bytes/s"] < 0)).sum())

print("\nFlow IAT Min 음수(620건) 중 Duration도 음수인 비율:")
print(df.loc[df["Flow IAT Min"] < 0, "Flow Duration"].lt(0).value_counts())

print("\nmin_seg_size_forward 음수값 분포:")
print(df.loc[df["min_seg_size_forward"] < 0, "min_seg_size_forward"].describe())

Header Length 음수값 분포:
count    2.300000e+01
mean    -3.100101e+08
std      4.618416e+08
min     -1.929350e+09
25%     -1.677705e+08
50%     -1.677704e+08
75%     -8.388512e+07
max     -8.388501e+07
Name: Fwd Header Length, dtype: float64

Duration 음수(17건)와 Packets/s·Bytes/s 음수 겹침:
Duration & Packets/s 동시: 17
Duration & Bytes/s 동시: 11

Flow IAT Min 음수(620건) 중 Duration도 음수인 비율:
Flow Duration
False    603
True      17
Name: count, dtype: int64

min_seg_size_forward 음수값 분포:
count          23.0
mean    -83885313.0
std             0.0
min     -83885313.0
25%     -83885313.0
50%     -83885313.0
75%     -83885313.0
max     -83885313.0
Name: min_seg_size_forward, dtype: float64


In [17]:
neg_dur_rows = df[df["Flow Duration"] < 0]
print("Duration<0 행의 Label 분포:")
print(neg_dur_rows["Label"].value_counts())

Duration<0 행의 Label 분포:
Label
BENIGN    17
Name: count, dtype: int64


- 데이터 자체 결함 정리

| # | 발견 | 건수 | 원인 | 처리 방침 | 근거 |
|---|---|---|---|---|---|
| 1 | Flow Duration = 0 → Bytes/s·Packets/s가 Inf/NaN | 471 | 0으로 나눈 값 (Inf 403+NaN 68=471=Duration 0건, 숫자로 증명) | 행 유지, Inf/NaN을 0으로 치환 + is_zero_duration 플래그 추가 | 이 중 DDoS 라벨 2건 존재 — 삭제 시 공격 데이터 손실 |
| 2 | Flow Duration < 0 | 17 | 타임스탬프 역전 추정 (Packets/s 17건 전부 겹침) | 행 삭제 | 시간은 논리적으로 음수 불가, 전부 BENIGN이라 라벨 편향 없음 |
| 3 | Header Length류 극단값 (Fwd/Bwd Header Length, min_seg_size_forward) | 최대 23 | 고정된 깨진 상수값 반복 (예: -83,885,313) — 계산 오버플로우 추정 | 행 삭제 | 동일 값 반복은 정상 분포가 아님, 규모 0.003%로 손실 미미 |
| 4 | Init_Win_bytes_forward/backward = -1 | 257,420 / 387,393 | TCP 윈도우 미관측 정상 표식 | 그대로 유지, 절대 삭제 안 함 | 이상치로 보고 지웠다가 전체 행 51% 손실난 v1 정제 실패 사례 있음 |
| 5 | 완전 중복 (Label 포함) | 29,568 | 캡처/생성 과정의 중복 | 행 삭제 | 동일 흐름 중복 기록은 학습/집계 왜곡 유발 |
| 6 | Flow IAT Min<0(603) / Fwd IAT Min<0(6), Duration<0과 안 겹치는 나머지 | 609 | 원인 미상 | 이번 범위에서 보존, README에 "추가 조사 필요"로 명시 | 전체의 0.08%, EDA 시간 대비 실익 낮음 |
| 7 | Fwd Header Length 중복 컬럼(.1) | — | pandas가 원본 중복 컬럼명에 자동 부여 | 값 동일 확인 후 하나만 남기고 삭제 | 같은 정보 이중 저장 불필요 |
| 8 | A_source 분석용 컬럼 | — | EDA 임시 컬럼 | 저장 전 삭제 | 원본 스키마에 없는 컬럼 |

### 3) 데이터 전처리 

In [18]:
# 1. 중복 컬럼 확인 후 정리
same = (df["Fwd Header Length"] == df["Fwd Header Length.1"]).all()
print("Fwd Header Length 중복 컬럼 값 동일:", same)
df = df.drop(columns=["Fwd Header Length.1"])

# 2. Duration=0: Inf/NaN 치환 + 플래그
df["is_zero_duration"] = df["Flow Duration"] == 0
df["Flow Bytes/s"] = df["Flow Bytes/s"].replace([np.inf, -np.inf], 0).fillna(0)
df["Flow Packets/s"] = df["Flow Packets/s"].replace([np.inf, -np.inf], 0)

# 3. Duration<0 행 삭제
before = len(df)
df = df[df["Flow Duration"] >= 0]
print("Duration<0 삭제:", before - len(df))

# 4. Header Length류 극단값 행 삭제
before = len(df)
broken_mask = (
    (df["Fwd Header Length"] < 0)
    | (df["Bwd Header Length"] < 0)
    | (df["min_seg_size_forward"] < 0)
)
df = df[~broken_mask]
print("Header Length 극단값 삭제:", before - len(df))

# 5. 완전 중복 삭제
before = len(df)
df = df.drop_duplicates()
print("완전 중복 삭제:", before - len(df))

# 6. 분석용 컬럼 제거
df = df.drop(columns=["A_source"])

print("\n최종 shape:", df.shape)

Fwd Header Length 중복 컬럼 값 동일: True


Duration<0 삭제: 17
Header Length 극단값 삭제: 23
완전 중복 삭제: 29568

최종 shape: (726055, 79)
